subsys/net/ip/connection.c

```diff
@@ -515,173 +523,228 @@ static enum net_verdict conn_raw_socket(struct net_pkt *pkt,
-			conn->cb, conn->user_data);
+	if (local->sll_ifindex != net_if_get_by_iface(pkt_iface)) {
+		return NET_CONTINUE;
+	}
 
-		raw_pkt = net_pkt_clone(pkt, CLONE_TIMEOUT);
-		if (!raw_pkt) {
-			net_stats_update_per_proto_drop(pkt_iface, proto);
-			NET_WARN("pkt cloning failed, pkt %p dropped", pkt);
-			return NET_DROP;
-		}
+	NET_DBG("[%p] raw match found cb %p ud %p", conn, conn->cb,
+		conn->user_data);
 
-		if (conn->cb(conn, raw_pkt, NULL, NULL, conn->user_data)
-		    == NET_DROP) {
-			net_stats_update_per_proto_drop(pkt_iface, proto);
-			net_pkt_unref(raw_pkt);
-		} else {
-			net_stats_update_per_proto_recv(pkt_iface, proto);
-		}
+	raw_pkt = net_pkt_clone(pkt, CLONE_TIMEOUT);
+	if (!raw_pkt) {
+		net_stats_update_per_proto_drop(pkt_iface, proto);
+		NET_WARN("pkt cloning failed, pkt %p dropped", pkt);
+		return NET_DROP;
+	}
 
-		return NET_OK;
+	if (conn->cb(conn, raw_pkt, NULL, NULL, conn->user_data) == NET_DROP) {
+		net_stats_update_per_proto_drop(pkt_iface, proto);
+		net_pkt_unref(raw_pkt);
+	} else {
+		net_stats_update_per_proto_recv(pkt_iface, proto);
 	}
 
-out:
-	return NET_CONTINUE;
+	return NET_OK;
 }
 
 enum net_verdict net_conn_input(struct net_pkt *pkt,
 				union net_ip_header *ip_hdr,
 				uint8_t proto,
 				union net_proto_header *proto_hdr)
 {
-	struct net_if *pkt_iface = net_pkt_iface(pkt);
-	struct net_conn *best_match = NULL;
-	bool is_mcast_pkt = false, mcast_pkt_delivered = false;
-	bool is_bcast_pkt = false;
-	bool raw_pkt_delivered = false;
-	bool raw_pkt_continue = false;
-	int16_t best_rank = -1;
-	struct net_conn *conn;
-	enum net_verdict ret;
-	uint16_t src_port;
-	uint16_t dst_port;
-
-	if (IS_ENABLED(CONFIG_NET_UDP) && proto == IPPROTO_UDP) {
-		src_port = proto_hdr->udp->src_port;
-		dst_port = proto_hdr->udp->dst_port;
-	} else if (IS_ENABLED(CONFIG_NET_TCP) && proto == IPPROTO_TCP) {
-		if (proto_hdr->tcp == NULL) {
+	uint8_t pkt_family = net_pkt_family(pkt);
+	uint16_t src_port = 0U, dst_port = 0U;
+
+	switch (pkt_family) {
+#if defined(CONFIG_NET_IP)
+	case AF_INET:
+	case AF_INET6:
+		switch (proto) {
+#if defined(CONFIG_NET_UDP)
+		case IPPROTO_UDP:
+			src_port = proto_hdr->udp->src_port;
+			dst_port = proto_hdr->udp->dst_port;
+			break;
+#endif
+#if defined(CONFIG_NET_TCP)
+		case IPPROTO_TCP:
+			if (proto_hdr->tcp == NULL) {
+				return NET_DROP;
+			}
+			src_port = proto_hdr->tcp->src_port;
+			dst_port = proto_hdr->tcp->dst_port;
+			break;
+#endif
+		default:
+			/* no-op */
+			break;
+		}
+		if (!conn_are_endpoints_valid(pkt, pkt_family, ip_hdr, src_port,
+					dst_port)) {
+			NET_DBG("Dropping invalid src/dst end-points packet");
 			return NET_DROP;
 		}
-
-		src_port = proto_hdr->tcp->src_port;
-		dst_port = proto_hdr->tcp->dst_port;
-	} else if (IS_ENABLED(CONFIG_NET_SOCKETS_PACKET)) {
-		if (net_pkt_family(pkt) != AF_PACKET ||
-		    (!IS_ENABLED(CONFIG_NET_SOCKETS_PACKET_DGRAM) &&
-		     proto != ETH_P_ALL && proto != IPPROTO_RAW)) {
+		break;
+#endif
```

Comment:
It'd be good to add comments to closing `#endif`, which `#if` do they correspond to, it's easy to get lost if they're nested (`/* defined(CONFIG_NET_IP) */` in this case for example).

Reply:
Fixed here and in all other places where I introduced the variable and where a nested #if exists.